### ISSUE TITLE
Feat: Ignore unlisted dependencies

### ISSUE THREAD

**Issue by mathieucaroff** [webpro/knip#381]
In my team, we use the alias feature of TypeScript, and as a result, we have thousands of unlisted dependencies in the report:

```
Unlisted dependencies (424)
@api                            cypress/e2e/common/ProxyConnectionRequestApproval/ProxyConnectionRequestApproval.spec.ts
@constants                      cypress/e2e/common/ProxyConnectionRequestApproval/ProxyConnectionRequestApproval.spec.ts
@helpers                        cypress/e2e/common/ProxyConnectionRequestApproval/ProxyConnectionRequestApproval.spec.ts
@selectors                      cypress/e2e/common/ProxyConnectionRequestApproval/ProxyConnectionRequestApproval.spec.ts
...
@api                            cypress/e2e/routes/targets/discovery/scanConfiguration/list/network/devices/list-network-devices.spec.ts
@constants                      cypress/e2e/routes/targets/discovery/scanConfiguration/list/network/devices/list-network-devices.spec.ts
@selectors                      cypress/e2e/routes/targets/discovery/scanConfiguration/list/utils.ts
@selectors                      cypress/support/helpers/feedback.ts
```

I wish there was a way to ignore unlisted dependencies by name, or better: by regex pattern. In this case, I could just ignore the dependency name pattern `^@[^/]+$` and keep using Knip happily :)

Related issue: https://github.com/webpro/knip/issues/159

**Comment by webpro** [comment]
Do you know about the `paths` config of Knip? https://knip.dev/reference/configuration/#paths

If not, would try that first :)

**Comment by webpro** [comment]
Oh, I just read Cypress (instead of TypeScript) first in the initial post. TypeScript `compilerOptions.paths` should work. I wonder if there's a bug. Any chance you could share more details or create a minimal reproduction?

**Comment by mathieucaroff** [comment]
Oww, well, here is how the files are organized:

```
adm
├── cypress
│   └── tsconfig.json
├── cypress.config.ts
├── cypress.env.json
└── tsconfig.json
```

It is in `adm/cypress/tsconfig.json` that the four alias are declared:

```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "types": ["cypress", "node", "cypress-file-upload"],
    "baseUrl": ".",
    "paths": {
      "@api": ["./support/api/index"],
      "@helpers": ["./support/helpers/index"],
      "@selectors": ["./support/selectors/index"],
      "@constants": ["./support/constants"]
    }
  },
  "include": ["e2e"]
}
```

I looked into `adm/cypress.config.ts` as well as `adm/cypress.env.json`, and saw nothing related to TypeScript nor `tsconfig`, so it appears that Cypress picks `adm/cypress/tsconfig.json` automatically by default.

**Comment by mathieucaroff** [comment]
While I'm at it, here's another problem I have with unlisted dependencies: Me and my team use the Ant Design component library and from time to time, we need to import file types which come from the internal component definitions. The thing is, the ant-design library is structured in a unique way, with an npm package declared for each base function component. These base component follow a naming convention: `rc-<component-name>` (rc stands for React Component). So we end up with this output in the `Unlisted dependencies` section of the knip report:

```
rc-picker/lib/interface         src/components/form/field/DateRangePicker/DateRangePickerField.tsx
rc-picker/lib/panels/TimePanel  src/components/form/field/DateRangePicker/DateRangePickerField.tsx
rc-tree-select/lib/TreeSelect   src/components/form/field/TreeSelectField/TreeSelectField-types.ts
rc-tree-select/lib/interface    src/components/form/field/TreeSelectField/TreeSelectField-types.ts
rc-select/lib/BaseSelect        src/routes/home/targets/_shared/Tags/form/_components/TagsCreateModal/TagsCreateModal-types.ts
rc-select/lib/BaseSelect        src/routes/home/targets/devices/form/tagsToDevicesAdd/TagsToDevicesAddForm-data.ts
rc-picker/lib/interface         src/routes/home/targets/discovery/onboarding/devices/_utils/table.tsx
rc-picker/lib/interface         src/routes/home/targets/discoveryOld/onboarding/devices/_utils/table.tsx
```

**Comment by webpro** [comment]
Perhaps Knip uses the root `tsconfig.json`, so it's missing out on the `paths` config. Some options I can think of include:

- If you're not in a monorepo, you can tell Knip to use the other `tsconfig.json` file: https://knip.dev/reference/cli/#--tsconfig-file
- You could copy over the `paths` config to the Knip config: https://knip.dev/reference/configuration/#paths

Please open separate tickets for separate issues.

**Comment by mathieucaroff** [comment]
I've tried changing the typescript configuration to the following:
```
typescript: {
    config: ['tsconfig.json', 'tsconfig.eslint.json', 'cypress/tsconfig.json'],
  }
```
but to no avail. I don't know why it didn't work.

Specifying the same `paths` as is present in `cypress/tsconfig.json` does work, so I'm going to keep the configuration like that.

**Comment by webpro** [comment]
> I've tried changing the typescript configuration to the following:
> 
> ```
> typescript: {
>     config: ['tsconfig.json', 'tsconfig.eslint.json', 'cypress/tsconfig.json'],
>   }
> ```
> 
> but to no avail. I don't know why it didn't work.

Not sure if we talk about the same thing, but what I meant something like this (adjust to your situation):

```sh
knip --tsConfig ./adm/cypress/tsconfig.json
```

> ~Specifying the same `paths` as is present in `cypress/tsconfig.json` does work, so I'm going to keep the configuration like that.~
> 
> Actually, I can't figure out which path I need to specify in the right-hand-side of the entries for `knip` not to report the imported files and values as unused.

The paths should be relative from the root where the `knip.json` also should be. So I'd expect something like this:

```jsonc
{
  "paths": {
    "@api": ["./adm/cypress/support/api/index"],
    // etc.
  }
}
```

You can use `knip --debug` for more verbose output.

**Comment by mathieucaroff** [comment]
Sorry, I should have shown where our package.json and knip.ts where in the file tree:

```
adm
├── cypress
│   └── tsconfig.json
├── cypress.config.ts
├── cypress.env.json
├── knip.ts
├── package.json
└── tsconfig.json
```

We do not have a package.json at the root of the repository since the server-side is in Python and has no use for JS packages.

**Comment by webpro** [comment]
I would need a minimal reproduction of the issue, please see https://knip.dev/guides/troubleshooting/#minimal-reproduction on how to create one.

**Comment by mathieucaroff** [comment]
Resolved by commit 82635fc89561e74c7598b1a0f42713a111af9c28
